lib/matplotlib/widgets.py

```diff
@@ -330,17 +335,34 @@ def __init__(self, ax, label, valmin, valmax, valinit=0.5, valfmt=None,
         self.valmin = valmin
         self.valmax = valmax
         self.valstep = valstep
+        if valstep is not None:
+            self._valsnap = np.asanyarray(valstep)
+            if self._valsnap.ndim == 0:
+                self._valsnap = np.arange(
+                    self.valmin, self.valmax + self._valsnap, self._valsnap
+                )
+            elif self._valsnap.ndim == 1:
+                self._valsnap = np.sort(self._valsnap)
+            else:
+                raise ValueError(
+                    "valstep can have 0 or 1 dimensions "
+                    f" but has {self.valstep.ndim}"
+                )
+        else:
+            self._valsnap = None
```

Comment:
I find it a bit awkward to have a public attribute and a private attribute that are decoupled but essentially hold the same information.

I propose this:

```suggestion
        if not isinstance(valstep, Number):
            self.valstep = np.asanyarray(valstep)
            if self.valsnap.ndim =! 1:
                raise ValueError(
                    f"valstep must have 1 dimension but has {self.valstep.ndim}"
                )
            self.valstep = np.sort(self.valstep)
```

We don't need to support passing 0-dim arrays - I think we generally don't throughout the library.

And adapt `_value_in_bounds()` accordingly.

Reply:
> I find it a bit awkward to have a public attribute and a private attribute that are decoupled but essentially hold the same information.

I was worried that if someone passes in a float for valstep and then later checked the value of `slider.valstep` they would be confused by the fact that it had turned into a numpy array. Or perhaps more worrisome is that that would be backwards incompatible for anyone accessing `slider.valstep` after slider creation. I don't imagine that's many people but I suppose it's non-zero. Does matplotlib have a policy on what counts as backwards incompatible?

Reply:
> We don't need to support passing 0-dim arrays - I think we generally don't throughout the library.

Ahh, I was doing that to check if it was a float vs an array vs something malformed. Where does `Number` come from? That seems like a better way to test this.

Reply:
> And adapt _value_in_bounds() accordingly.

Final question. Are you proposing I retain the old slider snapping code in addition to the new snapping to an array code? In your suggestion you eliminated the creation of an array when `valstep` is a float.

Reply:
`from numbers import Numbers` (python stdlib)

> Does matplotlib have a policy on what counts as backwards incompatible?

In general, we're very strict with backwards compatibility. Exceptions are case-by-case decisions.

> I was worried that if someone passes in a float for valstep and then later checked the value of `slider.valstep` they would be confused by the fact that it had turned into a numpy array. Or perhaps more worrisome is that that would be backwards incompatible for anyone accessing `slider.valstep` after slider creation.

I was conversely worried that somebody might set `slider.valstep`, but that would not update `_valsnap`. (Though I'm not sure if attribute changes currently work 100% as they do only take effect through `_update()` -> `_value_in_bounds()`.

For now I propose the following pragmatic approach:

- `self.valstep = valstep  # validation deferred to _value_in_bounds()` in `__init__`:
- Move all the conversion logic and checks to `_value_in_bounds`.

This is consistent with handling of all parameters. You pay a slight performance penalty, but that is negligable compared to a draw update.

Reply:
Done. Thanks for all the guidance on this by the way! It's very helpful and much appreciated.

Reply:
It turns out that `set_val` does not currently call `_value_in_bounds` so I added an optional `kwarg` `validate` to enable `set_val` to cal `value_in_bounds`. Defaults `False` for backwards compat.

Reply:
Thanks for the positive feedback! I'm sorry this is a bit of a hassle, but all the widget stuff seems a bit ad-hoc (that's been before my time :stuck_out_tongue_closed_eyes:).

I think it's an error to allow setting arbitrary values, that do not comply with bounds. In that sense, we don't need a back compat option (as a general rule, we still can fix bugs without back-comaptibility considerations). We should

- always coerce the value via `_value_in_bounds()`
- add an API change note for the changed behavior
- possibly add a warning to `set_val()` if `val` was modified by `value_in_bounds()`.

To keep things more manageable, I propose to defer changing `set_val()` to a separate PR.